internal/wasmldr/src/loader/compiled/tls.rs

```diff
@@ -45,47 +48,29 @@ impl WasiFile for Stream {
     }
 
     async fn get_fdflags(&mut self) -> Result<FdFlags, Error> {
-        Ok(FdFlags::APPEND | FdFlags::NONBLOCK)
+        Ok(FdFlags::APPEND)
     }
 
-    async fn read_vectored<'a>(&mut self, bufs: &mut [io::IoSliceMut<'a>]) -> Result<u64, Error> {
-        let (tcp, tls) = &mut *self.0.write().unwrap();
-
-        if tls.wants_read() {
-            tls.read_tls(tcp)
-                .map_err(|e| Error::io().context(e))
-                .context("could not read TLS ciphertext from TCP stream")?;
-
-            tls.process_new_packets()
-                .map_err(|e| Error::io().context(e))
-                .context("could not process new TLS packets")?;
+    async fn read_vectored<'a>(&mut self, bufs: &mut [IoSliceMut<'a>]) -> Result<u64, Error> {
+        if self.tls.wants_read() {
+            self.tls.read_tls(&mut self.tcp)?;
+            self.tls.process_new_packets()?;
         }
 
-        let n = tls
-            .reader()
-            .read_vectored(bufs)
-            .map_err(|e| match e.kind() {
-                io::ErrorKind::UnexpectedEof => Error::io().context("unexpected EOF"),
-                _ => Error::io(),
-            })
-            .context("could not read decrypted TLS ciphertext")?;
+        let n = match self.tls.reader().read_vectored(bufs) {
+            Err(e) if e.kind() == ErrorKind::UnexpectedEof => 0,
+            Err(e) => return Err(e.into()),
+            Ok(n) => n,
+        };
 
         Ok(n as u64)
     }
 
-    async fn write_vectored<'a>(&mut self, bufs: &[io::IoSlice<'a>]) -> Result<u64, Error> {
-        let (tcp, tls) = &mut *self.0.write().unwrap();
-
-        let n = tls
-            .writer()
-            .write_vectored(bufs)
-            .map_err(|e| Error::io().context(e))
-            .context("could not write plaintext to TLS buffer ciphertext")?;
+    async fn write_vectored<'a>(&mut self, bufs: &[IoSlice<'a>]) -> Result<u64, Error> {
+        let n = self.tls.writer().write_vectored(bufs)?;
```

Comment:
These need to be wrapped as was done before, because otherwise they will trigger a crash in wasmtime (it crashes on unknown errors). Good to have context as well.

Reply:
I think the issue is more complex than that. We definitely don't want to crash wasmtime. However, we need to be able to trigger the right wasm return values too. For example, nonblocking doesn't really work in part because we can't bubble up `ErrorKind::WouldBlock` properly. I removed the wrapping because I was concerned that our wrapping would "hide" the appropriate return values from wasmtime.

Reply:
But this is an error from rustls, not wasmtime, right?